core/indexer/common.go

```diff
@@ -361,102 +371,106 @@ func serializeAccountsHistory(accounts map[string]*AccountBalanceHistory) []byte
 		buffSlice = append(buffSlice, buff)
 	}
 
-	return buffSlice
+	return buffSlice, nil
 }
 
 func prepareSerializedDataForATransaction(
 	tx *Transaction,
 	selfShardID uint32,
-	isMBOfTxInDB bool,
-) (meta []byte, serializedData []byte) {
-	var err error
-	if isMBOfTxInDB {
-		if !isCrossShardDstMe(tx, selfShardID) || tx.Status == transaction.TxStatusInvalid.String() {
-			return
-		}
-		// update transaction
-		meta, serializedData = prepareTxUpdate(tx)
-	} else {
-		// insert transaction in database
-		meta = []byte(fmt.Sprintf(`{ "index" : { "_id" : "%s" } }%s`, tx.Hash, "\n"))
-		serializedData, err = json.Marshal(tx)
-		if err != nil {
-			log.Debug("indexer: marshal",
-				"error", "could not serialize transaction, will skip indexing",
-				"tx hash", tx.Hash)
-			return
-		}
-	}
-	return
-}
+	_ bool,
+) ([]byte, []byte, error) {
+	metaData := []byte(fmt.Sprintf(`{"update":{"_id":"%s", "_type": "_doc"}}%s`, tx.Hash, "\n"))
 
-func prepareSerializedAccountInfo(address string, account *AccountInfo) (meta []byte, serializedData []byte) {
-	var err error
-	meta = []byte(fmt.Sprintf(`{ "index" : { "_id" : "%s" } }%s`, address, "\n"))
-	serializedData, err = json.Marshal(account)
+	marshaledTx, err := json.Marshal(tx)
 	if err != nil {
 		log.Debug("indexer: marshal",
-			"error", "could not serialize account, will skip indexing",
-			"address", address)
-		return
+			"error", "could not serialize transaction, will skip indexing",
+			"tx hash", tx.Hash)
+		return nil, nil, err
 	}
-
-	return
-}
-
-func prepareSerializedAccountBalanceHistory(address string, account *AccountBalanceHistory) (meta []byte, serializedData []byte) {
-	var err error
-	meta = []byte(fmt.Sprintf(`{ "index" : { "_id" : "%s" } }%s`, address, "\n"))
-	serializedData, err = json.Marshal(account)
-	if err != nil {
-		log.Debug("indexer: marshal",
-			"error", "could not serialize account history entry, will skip indexing",
-			"address", address)
-		return
+	if !isCrossShardDstMe(tx, selfShardID) || tx.Status == transaction.TxStatusInvalid.String() {
+		serializedData :=
+			[]byte(fmt.Sprintf(`{"script":{"source":""},"upsert":%s}`,
```

Comment:
I think there are some fields (@miiu96 can add to this) that only the source shard has, right? Those should be added to the source field.

Reply:
no. based on the old implementation, on the `insert` part it only marshaled the transaction and indexed it. the same thing happens here.

Reply:
Discussed in call, it's ok like this 👍